File: planner/python/gtrain.py
### Train model using the GLean interface
import os
import sys
import time
import argparse
import time

import numpy as np

from gpvrnn import GPvrnn as model

if __name__ == '__main__':
    # Parsing arguments
    parser = argparse.ArgumentParser()
    parser.add_argument('config_file', help="configuration file for PV-RNN")
    args = parser.parse_args()  # will complain if no argument is given

    # Initialize PVRNN model
    rnn = model(config_file=args.config_file, task="training", epoch=-1) # resume training

    # Custom training loop
    bg_epochs = max(rnn.save_epochs // 10, 1)
    start_epoch = rnn.epoch
    start_time = time.time()
    loop_epochs = 0
    for i in range(start_epoch // bg_epochs, rnn.max_epochs // bg_epochs):
        loss = rnn.train(bg_epochs, greedy_train=False)
        loss_rec = loss["total_batch_reconstruction_loss"]
        loss_reg = loss["total_batch_regularization_loss"]
        loss_total = loss_rec+loss_reg
        print("Epoch %d  time %.0f / %.3f  loss_total %.7f  loss_rec %.7f  loss_reg %.7f" % ((i+1)*bg_epochs, time.time() - start_time, (time.time() - start_time)/((loop_epochs+1)*bg_epochs+1), loss_total, loss_rec, loss_reg))
        loop_epochs += 1

File: planner/gtrain.cpp
#include <glean.h>
#include <cnpy.h>
#include <chrono>

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Path to config file required as first argument" << std::endl;
        exit(EXIT_FAILURE);
    }

    GPvrnn* rnn = new GPvrnn(argv[1]); // using GLean interface

    // rnn->train(); // internal training routine

    /* Custom training routine */
    rnn->trainInitialize(-1); // resume training
    const int bg_epochs = std::max(rnn->save_epochs() / 10, 1);
    const int bg_start = rnn->current_epoch() / bg_epochs;
    const int bg_end = rnn->max_epochs() / bg_epochs;
    double rec_loss = 0.0;
    double reg_loss = 0.0;
    int etime, ctime;
    auto t0 = std::chrono::high_resolution_clock::now();
    for (int iterb = bg_start; iterb < bg_end; iterb++) {
        auto t1 = std::chrono::high_resolution_clock::now();
        rnn->train(bg_epochs, rec_loss, reg_loss, false); // Note: includes iotime
        auto t2 = std::chrono::high_resolution_clock::now();
        etime = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count();
        ctime = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t0).count();
        printf("Epoch %d  time %.0f / %.3f  loss_total %.7f  loss_rec %.7f  loss_reg %.7f\n", (iterb+1)*bg_epochs, ctime/1000.0, (etime/1000.0)/bg_epochs, rec_loss+reg_loss, rec_loss, reg_loss);
    }
}
